modules/dnn/src/cuda/math.cu

```diff
@@ -0,0 +1,243 @@
+// This file is part of OpenCV project.
+// It is subject to the license terms in the LICENSE file found in the top-level directory
+// of this distribution and at http://opencv.org/license.html.
+
+#include <cuda_runtime.h>
+
+#include "../cuda4dnn/csl/kernels.hpp"
+#include "../cuda4dnn/csl/kernel_utils.hpp"
+#include "../cuda4dnn/csl/span.hpp"
+#include "../cuda4dnn/csl/stream.hpp"
+
+namespace cv { namespace dnn { namespace cuda4dnn { namespace csl  { namespace kernels {
+
+    namespace detail {
+        template <class T> __device__ T abs(T val);
+        template <> __device__ float abs(float val) { return fabsf(val); }
+        template <> __device__ double abs(double val) { return fabs(val); }
+
+        template <class T> __device__ T exp(T val);
+        template <> __device__ float exp(float val) { return expf(val); }
+        template <> __device__ double exp(double val) { return ::exp(val); }
+
+        template <class T> __device__ T max(T x, T y);
+        template <> __device__ float max(float x, float y) { return fmaxf(x, y); }
+        template <> __device__ double max(double x, double y) { return fmax(x, y); }
+
+        template <class T> __device__ T min(T x, T y);
+        template <> __device__ float min(float x, float y) { return fminf(x, y); }
+        template <> __device__ double min(double x, double y) { return fmin(x, y); }
+
+        template <class T> __device__ T log1p(T val);
+        template <> __device__ float log1p(float val) { return log1pf(val); }
+        template <> __device__ double log1p(double val) { return ::log1p(val); }
+
+        template <class T> __device__ T log1pexp(T val);
+        template <> __device__ double log1pexp(double val) {
+            if (val <= -37)
+                return exp(val);
+            else if (-37 < val && val <= 18)
+                return log1p(exp(val));
+            else if (18 < val && val <= 33.3)
+                return val + exp(-val);
+            else
+                return val;
+        }
+        template <> __device__ float log1pexp(float val) { return log1pexp<double>(val); }
```

Comment:
Double precision performance is 2x slower compared to single precision performance on recent Nvidia GPUs and significantly slower for slightly old GPUs.

I picked up the double precision version from dlib. There are many libraries which use the same implementation but none of them provides an explanation on how they arrived at the approximations and the bounds. Hence, I haven't been able to write a single precision version.

The Caffe and OpenCV CPU implementations are using a numerically unstable version.

Reply:
Sorry I didn't answer this more clearly before.  I forget exactly how I selected these values specifically.  However, it's instructive to know that the double precision machine epsilon is about 1e-16.  If you add 1 to the machine epsilon you get back just 1.  Values larger than the epsilon, when added to 1, will give you a number > 1.  So when you have an expression like `x + exp(-x)`, this expression becomes numerically pointless when `exp(-x)` gets near the machine epsilon. It will yield exactly `x` once `exp(-x)` gets to the machine epsilon.  So you might as well just use `x` by itself.

So to be very specific, there isn't any point in doing `x + exp(-x)` for `x > 33.3` (approximately). You might as well just use `x`, which is why this value is used in the dlib code. The single precision epsilon is about 1e-7, so you could use 16 instead since `exp(-16)` is about 1e-7.

The other values could probably be kept the same for a single precision version of `log1pexp`.  But you should check by just comparing the output of a single precision version with the double precision version to make sure they are the same, or very close.

But in any case, don't cast to double on the GPU, because as you say, it's slow.

Reply:
This is what I have come up with:

```
float log1pexp(float x)
{
        if (x <= -20)
            return std::expf(x);
        else if (-20 < x && x <= 9.0)
            return std::log1pf(std::expf(x));
        else if(9.0 < x && x <= 14.6)
            return x + std::exp(-x);
        else
            return x;
```

Test Code: https://gist.github.com/YashasSamaga/f400d7f0fef56c485fc60b6180ead758
Test Output: https://gist.github.com/YashasSamaga/f6f9b658b4789e86d5835f344ba0f3d1

All the errors are just one representable number away from the correct answer except for three numbers.

Is this ok?

Reply:
Seems sensible to me